File: src/main/scala/dev/helium_build/record/NullRecorder.scala
package dev.helium_build.record

import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files

import cats.effect.Resource
import cats.implicits._
import dev.helium_build.build.BuildSchema
import dev.helium_build.conf.RepoConfig
import dev.helium_build.sdk.SdkInstallManager
import io.circe.Json
import org.apache.commons.io.FileUtils
import zio.{IO, RefM, ZIO, ZManaged}
import zio.blocking.Blocking

final class NullRecorder[R <: Blocking] private
(
  metadataCache: RefM[Map[String, Json]],

  cacheDir: File,
  protected override val sdkDir: File,
  protected override val schemaFile: File,
  override val sourcesDir: File,
  protected override val confDir: File,
) extends LiveRecorder[R] {

  override def sdkInstallManager: ZIO[R, Throwable, SdkInstallManager] =
    SdkInstallManager(cacheDir)

  override def recordTransientMetadata(path: String)(fetch: ZIO[R, Throwable, Json]): ZIO[R, Throwable, Json] =
    metadataCache.modify { metadata =>
      fetch.map { data =>
        (data, metadata + (path -> data))
      }
    }

  override def recordArtifact(path: String)(fetch: File => ZIO[R, Throwable, File]): ZIO[R, Throwable, File] =
    fetch(cacheDir)
}

object NullRecorder {

  def apply[R <: Blocking]
  (
    cacheDir: File,
    sdkDir: File,
    schemaFile: File,
    sourcesDir: File,
    confDir: File,
  ): ZManaged[R, Throwable, NullRecorder[R]] =
    ZManaged.fromEffect(
      for {
        metadataCache <- RefM.make(Map.empty[String, Json])
      } yield new NullRecorder[R](
        metadataCache,
        cacheDir = cacheDir,
        sdkDir = sdkDir,
        schemaFile = schemaFile,
        sourcesDir = sourcesDir,
        confDir = confDir,
      )
    )

}

File: src/Engine/Record/NullRecorder.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;
using Helium.Engine.Build;
using Helium.Engine.Cache;
using Helium.Engine.Conf;
using Helium.Sdks;
using Newtonsoft.Json.Linq;
using Nito.AsyncEx;

namespace Helium.Engine.Record
{
    internal class NullRecorder : LiveRecorder
    {
        public NullRecorder(string cacheDir, string sdkDir, string schemaFile, string sourcesDir, string confDir) {
            this.cacheDir = cacheDir;
            SdkDir = sdkDir;
            SchemaFile = schemaFile;
            SourcesDir = sourcesDir;
            ConfDir = confDir;
        }
        
        private readonly string cacheDir;
        private readonly ConcurrentDictionary<string, Task<JObject>> metadataCache = new ConcurrentDictionary<string, Task<JObject>>();

        public override string SourcesDir { get; }
        
        protected override string SchemaFile { get; }
        protected override string SdkDir { get; }
        protected override string ConfDir { get; }


        public override Task<string> RecordArtifact(string path, Func<string, Task<string>> fetch) =>
            fetch(cacheDir);

        public override Task<JObject> RecordTransientMetadata(string path, Func<Task<JObject>> fetch) =>
            metadataCache.GetOrAdd(path, _ => fetch());

        public override ISdkInstallManager CreateSdkInstaller() => new SdkInstallManager(cacheDir);

        public override async Task RecordMetadata() { }
    }
}